proxy/src/transparency/glue.rs

```diff
@@ -81,7 +81,7 @@ impl tower_h2::Body for HttpBody {
 
     fn is_end_stream(&self) -> bool {
         match *self {
-            HttpBody::Http1(_) => false,
+            HttpBody::Http1(ref b) => b.is_empty(),
```

Comment:
Sorry, this is the change I don't understand. In the change to the `hyper` crate that adds `b.is_empty()`, which is https://github.com/hyperium/hyper/commit/2f45d5394a2f8a49442ff4798a4b1651c079f0ff, there is this comment:

```rust
    /// This does **not** detect if the body stream may be at the end, or
    /// if the stream will not yield any chunks, in all cases. For instance,
    /// a streaming body using `chunked` encoding is not able to tell if
    /// there are more chunks immediately.
```

It seems like that comment is saying that we can't implement `is_end_stream()` in terms of `b.is_empty()`, but that's what we're doing here. Should there be a similar clarifying comment here?

Reply:
The `Body::is_end_stream` for tower-h2 is an optimization: if it returns true, h2 won't wait on another poll, and just signal end of stream right away. So, it's OK for this hint to be **false**, as it means it was just call `poll_data` again.

The comment from hyper is trying to clarify that `Body::is_empty()` means we already know definitively that there is no chunks in this body. However, consider if the server were to receive this request:

```
POST /foo HTTP/1.1
Host: localhost
Transfer-Encoding: chunked

0


```

Since hyper saw a `Transfer-Encoding` header, it can't say for sure that there is definitely no chunks. But once the user starts streaming the body, they'll receive a `None`, since it turned out there weren't any. To know otherwise would require us to do some buffering in the proxy.